src/persist-client/src/operators/shard_source.rs

```diff
@@ -298,117 +222,266 @@ where
+                // we can return immediately, as we could not have emitted any
+                // parts to fetch.
+                cap_set.downgrade(&[]);
+                return;
+            }
+            Either::Right(((subscription, as_of), _)) => {
+                (subscription, as_of)
+            }
+        };
+
+        let mut subscription = subscription.unwrap_or_else(|e| {
+            panic!(
+                "{}: {} cannot serve requested as_of {:?}: {:?}",
+                name_owned, shard_id, as_of, e
+            )
+        });
+
+        let mut batch_parts = vec![];
+
+        let mut lease_returner = subscription.clone_lease_returner();
+        let subscription_stream = async_stream::stream! {
+            // Eagerly yield the initial as_of. This makes sure that the output
+            // frontier of the `persist_source` closely tracks the `upper` frontier
+            // of the persist shard. It might be that the snapshot for `as_of` is
+            // not initially available yet, but this makes sure we already downgrade
+            // to it.
+            //
+            // Downstream consumers might rely on close frontier tracking for making
+            // progress. For example, the `persist_sink` needs to know the
+            // up-to-date upper of the output shard to make progress because it
+            // will only write out new data once it knows that earlier writes went
+            // through, including the initial downgrade of the shard upper to the
+            // `as_of`.
+            yield ListenEvent::Progress(as_of.clone());
+
+            let mut done = false;
+            while !done {
+                for event in subscription.next().await {
+                    if let ListenEvent::Progress(ref progress) = event {
+                        // If `until.less_equal(progress)`, it means that all subsequent batches will
+                        // contain only times greater or equal to `until`, which means they can be
+                        // dropped in their entirety. The current batch must be emitted, but we can
+                        // stop afterwards.
+                        if PartialOrder::less_equal(&until, progress) {
+                            done = true;
                         }
+                    }
+                    yield event;
+                }
+            }
+        };
+        tokio::pin!(subscription_stream);
 
-                        cap_set.downgrade(progress.iter());
-                        match progress.into_option() {
-                            Some(ts) => {
-                                current_ts = ts;
+        'emitting_parts:
+        loop {
+            tokio::select! {
+                // Order here is not required for correctness, but minimizes
+                // the work we need to do (e.g. check if the dataflow has been
+                // dropped before reading more data from CRDB)
+                biased;
```

Comment:
wanted to note that I personally find the `select!` based structure of this code to be pretty hard to reason about (for several reasons, [see here](https://blog.yoshuawuyts.com/futures-concurrency-3/#issues-with-futures-select) but also just things like code indentation). curious if the [stream merge](https://blog.yoshuawuyts.com/futures-concurrency-3/#concurrent-stream-processing-with-stream-merge) approach was considered and, if so, why it was rejected?

I don't want to throw a wrench in things on this PR; I'm happy to trust that y'all have thought about this carefully and support this merging as-is, but I'd like to keep our options open here to potentially structuring this code in a different way. and the biggest thing that I see that would make that hard is biased select, so I was wondering if we think it's necessary? if not, I'm inclined to go with non-biased select so it's one less stress if we decide to change things later

Reply:
I wanted the stream merge approach to work, but a few things make it tricky here --

* after the token is dropped, we want to continue reading from `completed_fetches` and only `completed_fetches`, but `Stream::merge` takes ownership of its inputs. AFAIK there isn't an easy way to only poll that stream after the main loop finishes
* haven't been able to think of a readable way of representing the subscribe vs flow control branching that's now expressed with the `if`-preconditions in `Stream::merge`

I'm definitely open to another approach here, but I haven't been able to make that particular one ergonomic

Reply:
cool, very glad to hear that it was considered! that's certainly enough for me

> after the token is dropped, we want to continue reading from completed_fetches and only completed_fetches, but Stream::merge takes ownership of its inputs. AFAIK there isn't an easy way to only poll that stream after the main loop finishes

I'd imagine that they continue to be polled, but they get thrown away. this is certainly related to (and possibly the same) question I asked about whether the biased select is necessary (for reasonably performance, I understand it's not necessary for correctness)

> haven't been able to think of a readable way of representing the subscribe vs flow control branching that's now expressed with the if-preconditions in Stream::merge

this could be totally off base, but my initial reaction is... don't :). we'd end up buffering parts to be fetched while the flow control was full, which certainly has some downsides. I can hand wave some ways to fix that if it becomes a problem, but would probably need to be prototyping this myself to have any real specifics

Reply:
One option, and actually one I prefer over `Stream::merge` sometimes, is to use https://docs.rs/futures/latest/futures/future/fn.select_all.html. With that, you need to produce `N` futures, each of which needs to return the same _type_ of value, which you can match, just like with `Stream::merge`. Could be easier to write than fully going with streams.

As for `if preconditions`, those can be built into the futures that are selected (but @pH14, if we stick with the `select!` approach, we definitely want at least some inline documentation on the various combinations of preconditions that are possible, so we make sure we don't end up with a `select!` that isn't actually doing anything

But overall, I think going with the `select!` here, solving the core problem, is fine, and we can refactor the exact _style_ of "selection" later!

Reply:
> As for if preconditions, those can be built into the futures that are selected

The bit I wasn't sure for a `select_all` or `Stream::merge` approach is how to ergonomically pass the precondition values in to the futures since the current/max inflight bytes is shared state right now.

Reply:
I played with this more, and I think there's a way we could potentially model the behavior here with nested `Stream::merge`s (one for reading from subscribe vs blocking on flow control feedback; one for that + token + completed fetches) but it quickly shaped up into a much more radical change than what's in this PR. so for now, sticking with the `select!`

Reply:
👍 thanks for looking into it!